Application/Security.js

```diff
@@ -100,18 +100,117 @@ angular.module('shared')
           }
         );
       },
+      oAuth2Authorize: function(client_id, redirect_uri, response_type, scope, user, callback) {
+        var self = this;
+        
+        switch (response_type) {
```

Comment:
Here's how I think this function could be rewritten, if you rewrite the `APISupport` factory to return promises instead (either using `$q.defer()` or just returning the `$http()` call).

```
oAuth2Authorize: function (client_id, redirect_uri, response_Type, scope, user, callback) {
  var self = this;

  if (!response_type || response_type !== 'code') {
    // Log an error? Not sure if silently failing is what we want here
    return;
  }

  APISupport.oAuth2Authorize({
    client_id: client_id,
    redirect_uri: redirect_uri,
    response_type: 'code',
    scope: scope,
    user_id: user
  })
  .then(function (authCode) {
    if (!authCode.authorizationCode) {
      console.log("Error authorizing application with oAuth2 server");
      return;
    }

    return APISupport.oAuth2Token({
      code: authCode.authorizationCode,
      client_id: client_id,
      client_secret: null,
      grant_type: 'authorization_code',
      redirect_uri: authCode.redirectUri
    })
  })
  .then(function (accessToken) {
    accessToken.id = accessToken.accessToken;
    LoopBackAuth.setUser(accessToken.id, null, false);
    LoopBackAuth.save();

    // I would also argue that these should all just be an object with named properties, that way it's
    // way easier to tell what is what, rather than thinking "why is the 3rd argument an empty object..."
    self.initLogin(accessToken, {
      rememberme:true
    }, {}, {
      loginRedirect: accessToken.client.redirectUri
    }, user);
  })
  .catch(function (err) {
    console.log('Error during oAuth2Authorize', err);
  });
}
```